### PR TITLE
tracee-ebpf: add BTFhub support

### PR DESCRIPTION
**Pull request by rafaeldtinoco**
tracee-ebpf: add support to BTFhub

    Enable BTFHUB integration by compiling tracee-ebpf with:

        $ STATIC=1 BTFHUB=1 make -f Makefile.one tracee-ebpf

    and executing in a BTFHUB supported environment:

        $ sudo ./dist/tracee-ebpf --debug --trace 'event!=sched*'
        OSInfo: KERNEL_RELEASE: 5.4.0-91-generic
        OSInfo: ARCH: x86_64
        OSInfo: VERSION: "18.04.6 LTS (Bionic Beaver)"
        OSInfo: ID: ubuntu
        OSInfo: ID_LIKE: debian
        OSInfo: PRETTY_NAME: "Ubuntu 18.04.6 LTS"
        OSInfo: VERSION_ID: "18.04"
        OSInfo: VERSION_CODENAME: bionic
        BTF: bpfenv = false, btfenv = false, vmlinux = false
        BTF: using BTF file from embedded btfhub: /tmp/tracee/tracee.btf
        unpacked CO:RE bpf object file into memory
        TIME             UID    COMM             PID     TID...

    Note: if you're willing to execute btfhub powered tracee-ebpf inside a
          container then make sure to run your container using the
          Makefile.docker "alpine-tracee-shell" or "ubuntu-tracee-shell"
          targets, as they bind mount /etc/os-release from host into the
          container.

----

1. I'm not adding to workflow tests on purpose. I think we should wait until this feature is totally integrated to the release process in order to start testing it on PRs.

2. I'll update building documents (docs/) once Makefile.one becomes standard. Then I'll "teach" how to build tracee-ebpf with embedded BTF files and supported environments from BTFhub, etc.

### CODE REVIEW

**Comment by rafaeldtinoco** [comment]
I'd rather wait PR https://github.com/aquasecurity/tracee/pull/1227 in order to continue this (as BTFHUB=1 option needs Makefile to have correct target dependencies).

**Comment by yanivagman** [comment]
To track previous comments:
Previous PR #1125

File: tracee-ebpf/main.go
Comment (yanivagman) [inline tracee-ebpf/main.go:1197]: It might be a problem to use this API of OSInfo in tracee container (alpine based) as we will get the wrong distro release info (aquasecurity/libbpfgo#92)

Comment (rafaeldtinoco) [reply tracee-ebpf/main.go:1197]: Yes, we're mitigating this by doing: https://github.com/aquasecurity/tracee/pull/1227/commits/76883662956efb30e9d7d22b00d9a5a719492624#diff-9622619b5207ed1b3ace8ccab271f0f000e5271e1d6412a5b598a3171f5f1080R47

**Comment by rafaeldtinoco** [comment]
@yanivagman,

Obviously I'm only including the feature in the new Makefile (Makefile.one) as one of the reasons I refactored it was to include the BTFhub support in more organized way. 

```
$ STATIC=1 BTFHUB=1 make -f Makefile.one
```

Shall do the trick.

**Comment by rafaeldtinoco** [comment]
I'm not adding to workflow tests on purpose. I think we should wait until this feature is totally integrated to the release process in order to start testing it on PRs.

File: 3rdparty/btfhub.sh
Comment (rafaeldtinoco) [inline 3rdparty/btfhub.sh:67]: Im excluding those here because they already support embedded BTF and I want to include only BTFs for "really needed" environments.

File: Makefile.one
Comment (rafaeldtinoco) [inline Makefile.one:368]: @danielpacak we "forgot" after we changed things to /cmd/. We might need to revisit the Makefile each time things are moved around as, for example, its likely we should also list ./pkg/* (this way whenever ./pkg/*.go changes the target is rebuilt).

File: Makefile.one
Comment (rafaeldtinoco) [inline Makefile.one:440]: @danielpacak and here as well.

File: embedded-btfhub.go
Comment (yanivagman) [inline embedded-btfhub.go:14]: So, this file is the same as embedded-ebpf.go except this line, right? Maybe we can just add this line to the other file, embedding an empty dir (which we can create in the makefile) and avoid adding this new file?

**Review by yanivagman — COMMENTED** [review]
Awesome work @rafaeldtinoco!

File: cmd/tracee-ebpf/main.go
Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/main.go:228]: These lines are about to be changed by another PR I have ready (dropping the getBPFOjectPath() building logic and having bpf-nocore built and installed by Makefile (or docker builder)).

Mind if we do that change in that PR ? I can then have 1 commit cleaning this and the 2nd commit dropping the build logic to the builder.

File: embedded-btfhub.go
Comment (rafaeldtinoco) [reply embedded-btfhub.go:14]: Yes, that was my initial idea in the other PR that was closed. There was a `touch ./dist/btfhub/place-holder`, maybe you recall, as empty dirs can't be embedded. I do agree that either having:

1. ./dist/btfhub/place-holder
2. ./dist/btfhub/{ubuntu,centos,fedora}

is better than this embedded-btfhub.go logic. Let me go back to initial idea then.

File: cmd/tracee-ebpf/main.go
Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/main.go:228]: Nevermind, I'll improve that in this PR.. this way its cleaner for the next one.

Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/main.go]: I moved that extensive logic into a function called `dealWithGivenEnv()`, and organized the decision taking logic into 5 ordered steps, enumerating where they are.

The `if debug {}` snippets make this look more complex than it really is, but I think, with steps enumerated, it got easier to understand/read. 

Of course, there is always space for improvement AND it will be easier to simplify this once "getBPFOjectPath()" is removed (my next PRs, almost ready).

Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/main.go]: BTW if you think there is a better way of doing this (with goto's) please could you recommend it as a snippet here ? Or if you prefer to do it later, ok too. Your call.

File: cmd/tracee-ebpf/main.go
Comment (yanivagman) [inline cmd/tracee-ebpf/main.go:142]: Isn't `dealWithGivenEnv()` a too wide name?
Maybe `prepareBpfObject()`?

File: cmd/tracee-ebpf/main.go
Comment (yanivagman) [inline cmd/tracee-ebpf/main.go:359]: You can set a label here for goto, e.g. `out:`

File: cmd/tracee-ebpf/main.go
Comment (yanivagman) [inline cmd/tracee-ebpf/main.go:316]: goto out (and remove "else" and indentation below)

File: cmd/tracee-ebpf/main.go
Comment (yanivagman) [inline cmd/tracee-ebpf/main.go:333]: goto out (and remove "else" and indentation below)

File: cmd/tracee-ebpf/main.go
Comment (yanivagman) [inline cmd/tracee-ebpf/main.go:346]: goto out (and remove "else" and indentation below)

**Review by danielpacak — APPROVED** [review]
I'm still catching up with CO:RE, BTF, and btfhub, but I approve based on the test that I run in the Vagrant VM:

```
BTFHUB=1 make -f Makefile.one
sudo ./dist/tracee-ebpf
```

And then validating that ./dist/btfhub/ubuntu/20.04/x86_64/5.4.0-90-generic.btf is copied from the embed.FS to /tmp/tracee/tracee.btf when we run tracee-ebpf

File: cmd/tracee-ebpf/main.go
Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/main.go]: Okay, the overall result is better indeed. By using goto's, like you said, we're making it easier to read and also removing a bunch of indentation (which I prefer as well). Nice tip.

**Comment by rafaeldtinoco** [comment]
@yanivagman addressed your comments.

**Comment by rafaeldtinoco** [comment]
Alright, working on this (the build failure). Got a meeting now and will fix it right after.

**Comment by rafaeldtinoco** [comment]
@yanivagman could u +1 decent changes before I merge ?

**Review by yanivagman — APPROVED** [review]

File: cmd/tracee-ebpf/main.go
Comment (itaysk) [inline cmd/tracee-ebpf/main.go:296]: I think this flow can be greatly simplified as follows:

```
if bpfenv {
  config.bpf = bpfenv
} else {
  config.bpf =  unpack
}
if btfenv {
  config.btf = btfenv
} else if vmlinux {
  config.btf = vmlinux
}
```

WDYT?

File: 3rdparty/btfhub.sh
Comment (yanivagman) [reply 3rdparty/btfhub.sh:69]: Talking about tools which are not installed OOB, pkg-config (added in the makefile PR) is another such a tool.

File: Makefile.one
Comment (rafaeldtinoco) [reply Makefile.one:414]: I did not understand what you mean here. What was innocently slipped ? The script ? The ".ONESHELL" special target ?

File: 3rdparty/btfhub.sh
Comment (rafaeldtinoco) [reply 3rdparty/btfhub.sh:1]: I'm installing it in our "tracee-make" container (but not making it the default). The reason for me to install is to give a better experience for "tracee-make" container shell users (set -o vi mode is more complete, for example, among others). 

For the script, I'm using "bash"-isms on it. I can try to use /bin/sh by default, but, since we're installing it I thought it would be okay to for the script as well.

File: 3rdparty/btfhub.sh
Comment (rafaeldtinoco) [reply 3rdparty/btfhub.sh:2]: Yep, I agree.

File: 3rdparty/btfhub.sh
Comment (rafaeldtinoco) [reply 3rdparty/btfhub.sh:100]: They are both ignored by .gitignore. But I'm not sure I follow this comment, could you clarify it ?

File: cmd/tracee-ebpf/main.go
Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/main.go:265]: Yep, I'll change it. It was suggested by Yaniv in another review, I thought about it, but forgot to change Bpf to BPF =(. Will do. (And maybe we should start enforcing style together with gofmt in workflows).

File: cmd/tracee-ebpf/main.go
Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/main.go:296]: Definitely. I knew this had space for improvement, I'll improve it by using this suggestion.

Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/main.go:296]: @danielpacak like we've spoken, would you like to simplify this and get into the BPF and BTF loading logic ?

File: 3rdparty/btfhub.sh
Comment (itaysk) [reply 3rdparty/btfhub.sh:100]: these dirs holds the output of the script right? so I don't think they should be committed to git, and expected to come from the git clone, but created by the script. similar to how we treat `dist` in tracee's build scripts

File: cmd/tracee-ebpf/main.go
Comment (danielpacak) [reply cmd/tracee-ebpf/main.go:296]: Sure, I've created #1275 to work on that.

File: 3rdparty/btfhub.sh
Comment (itaysk) [reply 3rdparty/btfhub.sh:80]: Thanks for explaining. I think there should be an easy way to get just the relevant kernels from btfhub, but this is a future enhancement.

File: Makefile.one
Comment (itaysk) [reply Makefile.one:414]: the `.ONESHELL`. it changes the way the makefile is processed. I know it might be a good change, but still not the normal behavior. So I'm suggesting to either reconsider it (if it's just for one target), or to move it to the top where it's visible to readers

Comment (rafaeldtinoco) [reply Makefile.one:414]: Hum, I'm using it a lot, specially in the fixes I made for this Makefile in the PR of dropping the non core BPF build. There is a commit there fixing a bunch of stuff in this Makefile. I think we can drop ONE SHELL entirely there to make the Makefile not to rely in that feature (which would be ok I suppose)

File: cmd/tracee-ebpf/main.go
Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/main.go:296]: Thanks! Marking this as resolved (per new issue created).

File: Makefile.one
Comment (rafaeldtinoco) [reply Makefile.one:414]: Okay, so.. I'm using the .ONESHELL feature *a lot* now. For:

1. .check_%
2. .checklib_%
3. .checkver_$(CMD_CLANG)
4. .checkver_$(CMD_GO)
5. btfhub

It allows us to have more complex shell scripting within the Makefile, which is great for features to check things (since we're checking a bunch of stuff in the makefile now). I'd rather keep it and move it to the top, like you said. Doing it in a "makefileadjusts" PR.

Comment (rafaeldtinoco) [reply Makefile.one:414]: https://github.com/aquasecurity/tracee/pull/1276

File: 3rdparty/btfhub.sh
Comment (rafaeldtinoco) [reply 3rdparty/btfhub.sh:69]: I added rsync as its a better tool (and small) for the job. pkg-config is part of the distros currently and does not need to be installed in the builders (already present). I'm checking the existence of all commands being used in the Makefile at the PR: https://github.com/aquasecurity/tracee/pull/1276

File: cmd/tracee-ebpf/main.go
Comment (rafaeldtinoco) [reply cmd/tracee-ebpf/main.go:265]: Being done at PR mentioned at https://github.com/aquasecurity/tracee/pull/1226#discussion_r770726554

File: 3rdparty/btfhub.sh
Comment (rafaeldtinoco) [reply 3rdparty/btfhub.sh:2]: https://github.com/aquasecurity/tracee/pull/1276/commits/a8b8507630a3d7af817675684522c04b4814b1e3

File: 3rdparty/btfhub.sh
Comment (rafaeldtinoco) [reply 3rdparty/btfhub.sh:100]: https://github.com/aquasecurity/btfhub/issues/25

**Pull request merged**
